File: stasis-core/src/de/boxxit/stasis/serializer/MethodResultSerializer.java
package de.boxxit.stasis.serializer;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.Serializer;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import de.boxxit.stasis.MethodResult;

/**
 * User: Christian Fruth
 */
public class MethodResultSerializer extends Serializer<MethodResult>
{
	public MethodResultSerializer()
	{
	}

	@Override
	public void write(Kryo kryo, Output output, MethodResult object)
	{
		output.writeInt(object.getType().ordinal(), true);

		switch (object.getType())
		{
			case Void:
				break;

			case Value:
			case Exception:
				kryo.writeClassAndObject(output, object.getResult());
				break;
		}
	}

	@Override
	public MethodResult read(Kryo kryo, Input input, Class<MethodResult> type)
	{
		MethodResult methodResult = new MethodResult();
		kryo.reference(methodResult);

		MethodResult.Type resultType = MethodResult.Type.values()[input.readInt(true)];
		Object result = null;

		switch (resultType)
		{
			case Void:
				break;

			case Value:
			case Exception:
				result = kryo.readClassAndObject(input);
				break;
		}

		methodResult.setType(resultType);
		methodResult.setResult(result);

		return methodResult;
	}
}

File: stasis-ios/Stasis/MethodResultSerializer.m
//
//  MethodResultSerializer.m
//  Stasis
//
//  Created by Christian Fruth on 24.10.14.
//  Copyright (c) 2014 Boxx IT Solutions GmbH. All rights reserved.
//

#import "MethodResultSerializer.h"
#import "MethodResult.h"
#import "Kryo/Kryo.h"

@implementation MethodResultSerializer

- (BOOL)acceptsNull
{
	return NO;
}

- (void)write:(Kryo *)kryo value:(id)value to:(KryoOutput *)output
{
	MethodResult *methodResult = value;
	
	[output writeInt:(SInt32)methodResult.type optimizePositive:YES];
	
	switch (methodResult.type)
	{
		case MethodResultVoid:
			break;
			
		case MethodResultValue:
		case MethodResultException:
			[kryo writeClassAndObject:methodResult.result to:output];
			break;
	}
}

- (id)read:(Kryo *)kryo withClass:(Class)type from:(KryoInput *)input
{
	
	MethodResult *methodResult = [MethodResult new];
	[kryo reference:methodResult];
	
	MethodResultType resultType = (MethodResultType)[input readIntOptimizePositive:YES];
	NSObject *result = nil;
	
	switch (resultType)
	{
		case MethodResultVoid:
			break;
			
		case MethodResultValue:
		case MethodResultException:
			result = [kryo readClassAndObject:input];
			break;
	}
	
	methodResult.type = resultType;
	methodResult.result = result;
	
	return methodResult;
}

@end
